content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
+
+### Call-chain clustering
+[Call-chain clustering (C3)](https://research.fb.com/publications/optimizing-function-placement-for-large-scale-data-center-applications-2/) is a recent function layout optimization developed at Facebook Research that addresses the issue with using undirected call graphs. We do not discuss its internal details in this post, but a simplified way of thinking about C3 is that it is PH but with a directed call graph. In this way, it improves spatial locality by considering caller-callee relations.
+
+## Codestitcher
+
+This brings us to Codestitcher, an optimizer that integrates basic block reordering and function reordering to improve upon the PH model. Below is a high-level overview of the steps taken in Codestitcher to generate the optimized code layout, given an input program.
+
+1. Compile the program to LLVM bitcode. This serves as the baseline program. As the program is built, symbols are emitted that allow us to generate a CFG for the program.
+2. Profile the program, generating weights for CFG edges.
+3. Using the weighted CFG, perform basic block chaining, ordering basic blocks in a way to minimize jumps.
+4. Perform hierarchical code collocation, ordering these basic block chains to form higher-level layouts. Here, the *d*-close partial layout algorithm works to build layouts to maximize spatial locality.
+
+There is not much to say about step 1, so let us explore steps 2, 3, and 4 in more detail.
+
+### Profiling
+In order to build our weighted CFG for layout analysis, we need to collect data about "control flow transfers" (when control transfers between CFG nodes). Luckily, most Intel CPUs come with a "last branch record", or LBR--- storing recently executed instructions such as jumps, branches, and calls. With the Linux `perf` tool, this data can be gathered and used to generate a weighted CFG for optimizing program layout.
+
+### Basic Block Chaining
+As we reorder blocks, we will need to insert additional jumps to preserve the control flow. Therefore, we wish to order basic blocks in a way that minimizes these unconditional jumps. This is to prevent unnecessary instruction pointer moves during the execution of the program. We define a basic block chain as a directed sequence of basic blocks where no unconditional jumps occur except possibly in the last block. Generating a maximum-cardinality set of these chains would constitute the fewest number of unconditional jumps required. The authors denote this problem the *fall-through maximization problem*. This problem is the same as the maximum path cover problem, which is **NP-hard**. PH uses a greedy heuristic to approach this but does not provide a theoretical guarantee of how close to optimal the generated solution is.
```

Comment:
> As we reorder blocks, we will need to insert additional jumps to preserve the control flow. Therefore, we wish to order basic blocks in a way that minimizes these unconditional jumps. This is to prevent unnecessary instruction pointer moves during the execution of the program.

This is part of the problem setup for _any_ code layout technique—so if you need to make this point, make it earlier.

Reply:
> where no unconditional jumps occur except possibly in the last block.

In fact, no jumps at all (conditional or unconditional) can appear anywhere but at the end.